cogs/cryochamber.py

```diff
@@ -0,0 +1,72 @@
+import re
+
+from discord import Embed, Colour
+from discord.ext.commands import Cog, CommandError, MessageConverter, TextChannelConverter, group
+
+
+class Cryochamber(Cog):
+    def __init__(self, _):
+        self.ongoing_preservations = set()
+
+    @group(invoke_without_command=True)
+    async def preserve(self, context, *, arg):
+        get_channel = TextChannelConverter().convert
+        get_message = MessageConverter().convert
+        exp = r'(?P<source_specifier>.*?) in (?P<destination_channel_specifier>.*?)'
+        command_match = re.fullmatch(exp, arg)
+
+        if command_match is None:
+            raise CommandError('wrong syntax: ' + arg)
+
+        dest_channel = await get_channel(context,
+                                         command_match.group('destination_channel_specifier'))
+
+        if not dest_channel.permissions_for(context.author).manage_messages:
+            raise MissingPermissions(['manage_messages'])
+
+        exp = r'(?:(?P<temporal_modifier>future|no more) )' \
+              r'?pins from (?P<source_channel_specifier>.*?)'
+        source_channel_match = re.fullmatch(exp,
+                                            command_match.group('source_specifier'))
+
+        if source_channel_match is None:
+            source_message = await get_message(context, command_match.group('source_specifier'))
+            await self._preserve_message(source_message, dest_channel)
+        else:
+            src_channel = await get_channel(context,
+                                            source_channel_match.group('source_channel_specifier'))
+            temporal_modifier = source_channel_match.group('temporal_modifier')
+
+            if temporal_modifier is None:
+                for pin in sorted(await src_channel.pins(), key=lambda message: message.created_at):
+                    await self._preserve_message(pin, dest_channel)
+            elif temporal_modifier == 'future':
+                self.ongoing_preservations.add((src_channel.name, dest_channel.name))
+            elif temporal_modifier == 'no more':
+                self.ongoing_preservations.remove((src_channel.name, dest_channel.name))
+            else:
+                raise ValueError('unknown temporal modifier: ' + temporal_modifier)
+
+    @preserve.command()
+    async def what(self, context):
```

Comment:
Documentation for the `what` subcommand can go as a docstring here, and would show up next to the subcommand.

But is this meant to be a subcommand? It would be executed by typing `?preserve what` *argument1* *argument2*. Some usage examples in the docstring would help again.

Reply:
I wasn't the one who wrote ?preserve what but I explained it as best I could